File: Difficulty: Medium/Koko Eating Bananas/koko-eating-bananas.cpp
class Solution {
  public:
    // Function to check if Koko can finish all bananas at speed 'mid' within 'k' hours
    bool finish(vector<int>& arr, int mid, int k) {
        int hours = 0;

        for (int i = 0; i < arr.size(); i++) {
            hours += arr[i] / mid;         // Add full chunks
            if (arr[i] % mid != 0) {
                hours++;                   // Add 1 hour for leftover bananas
            }
        }

        return hours <= k;  // Return true if total hours needed is within limit
    }
    
    int kokoEat(vector<int>& arr, int k) {
        int low    = 1;
        int high   = *max_element(arr.begin(), arr.end()); // Max bananas in a pile
        int result = 0;

        while (low <= high) {
            int mid = (low + high) / 2;  // Mid speed

            if (finish(arr, mid, k)) {
                result = mid;        // Try smaller speed
                high   = mid - 1;
            } else {
                low = mid + 1;       // Need to try faster speed
            }
        }

        return result;  // Minimum speed at which Koko can eat all bananas
    }
};

File: Difficulty: Medium/Koko Eating Bananas/koko-eating-bananas.java
class Solution {
    // Helper function to check if Koko can finish all bananas at speed 'mid' within 'k' hours
    private boolean canFinish(int[] arr, int mid, int k) {
        int hours = 0;

        for (int bananas : arr) {
            hours += bananas / mid;
            if (bananas % mid != 0) {
                hours++; // Add extra hour if leftover bananas exist
            }
        }

        return hours <= k;
    }

    public int kokoEat(int[] arr, int k) {
        int low  = 1;
        int high = Integer.MIN_VALUE;

        // Find the maximum number in the array
        for (int bananas : arr) {
            high = Math.max(high, bananas);
        }

        int result = 0;

        while (low <= high) {
            int mid = (low + high) / 2;

            if (canFinish(arr, mid, k)) {
                result = mid;       // Try smaller speed
                high   = mid - 1;
            } else {
                low = mid + 1;      // Try higher speed
            }
        }

        return result;
    }
}
